File: new/common/src/main/java/net/shiroha233/roadweaver/features/path/decoration/base/Decoration.java
package net.shiroha233.roadweaver.features.path.decoration.base;

import net.minecraft.core.BlockPos;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.tags.BlockTags;
import net.minecraft.world.level.levelgen.Heightmap;
import net.shiroha233.roadweaver.features.path.decoration.compat.RoadFeatureCompat;

/**
 * 装饰基类：提供位置计算和放置检查。
 */
public abstract class Decoration {
    private BlockPos placePos;
    private final WorldGenLevel world;

    public Decoration(BlockPos placePos, WorldGenLevel world) {
        this.placePos = placePos;
        this.world = world;
    }

    public abstract void place();

    public BlockPos computeSurfacePos() {
        return new BlockPos(placePos.getX(), world.getHeight(Heightmap.Types.WORLD_SURFACE_WG, placePos.getX(), placePos.getZ()), placePos.getZ());
    }

    public boolean isPlaceAllowedAt(BlockPos surfacePos) {
        BlockState below = world.getBlockState(surfacePos.below());
        boolean belowInvalid = below.is(Blocks.WATER) || below.is(Blocks.LAVA) || below.is(BlockTags.LOGS) || RoadFeatureCompat.dontPlaceHere(below.getBlock());
        return !belowInvalid;
    }

    protected final boolean placeAllowed() {
        BlockPos surfacePos = computeSurfacePos();
        this.placePos = surfacePos;
        return isPlaceAllowedAt(surfacePos);
    }

    public BlockPos getPos() { return placePos; }

    public WorldGenLevel getWorld() { return world; }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/features/path/decoration/base/Decoration.kt
package net.shiroha233.roadweaver.features.path.decoration.base

import net.minecraft.core.BlockPos
import net.minecraft.tags.BlockTags
import net.minecraft.world.level.WorldGenLevel
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.levelgen.Heightmap
import net.shiroha233.roadweaver.features.path.decoration.compat.RoadFeatureCompat

abstract class Decoration(
    private var placePos: BlockPos,
    private val world: WorldGenLevel
) {
    abstract fun place()

    fun computeSurfacePos(): BlockPos {
        return BlockPos(
            placePos.x,
            world.getHeight(Heightmap.Types.WORLD_SURFACE_WG, placePos.x, placePos.z),
            placePos.z
        )
    }

    fun isPlaceAllowedAt(surfacePos: BlockPos): Boolean {
        val below: BlockState = world.getBlockState(surfacePos.below())
        val belowInvalid = below.`is`(Blocks.WATER) ||
            below.`is`(Blocks.LAVA) ||
            below.`is`(BlockTags.LOGS) ||
            RoadFeatureCompat.dontPlaceHere(below.block)
        return !belowInvalid
    }

    protected fun placeAllowed(): Boolean {
        val surfacePos = computeSurfacePos()
        placePos = surfacePos
        return isPlaceAllowedAt(surfacePos)
    }

    fun getPos(): BlockPos = placePos

    fun getWorld(): WorldGenLevel = world
}
